File: src/main/java/pkg/exoad/swiggie/stx/SwPair.java
package pkg.exoad.swiggie.stx;

import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.lang.ref.WeakReference;

@Accessors(fluent = true) @Getter @Setter public class SwPair<A, B>
{
	public static <A, B> SwPair<A, B> acquire(
			A first,
			B second
	)
	{
		return new SwPair<>(
				first,
				second
		);
	}

	public static <A, B> WeakReference<SwPair<A, B>> acquireWeak(
			A first,
			B second
	)
	{
		return new WeakReference<>(new SwPair<>(
				first,
				second
		));
	}

	private A first;
	private B second;

	private SwPair(
			A first,
			B second
	)
	{
		this.first = first;
		this.second = second;
	}
}

File: src/main/java/pkg/exoad/swiggie/stx/SwPair.kt
package pkg.exoad.swiggie.stx

import pkg.exoad.swiggie.intf.SwSwappable
import pkg.exoad.swiggie.tools.defs.SwUnsafe
import java.lang.ref.WeakReference

class SwPair<A:Any , B> private constructor(private var first:A , private var second:B):SwSwappable
{
	fun first():A
	{
		return first
	}
	
	fun second():B
	{
		return second
	}
	
	fun first(first:A)
	{
		this.first=first
	}
	
	fun second(second:B)
	{
		this.second=second
	}
	
	@SwUnsafe(reason="Unable to perform runtime type checking for generics on a Pair!")
	@Suppress("UNCHECKED_CAST")
	override fun swap()
	{
		val temp:Any=first
		first=second as A
		second=temp as B
	}
	
	companion object
	{
		fun <A:Any , B> acquire(first:A , second:B):SwPair<A , B>
		{
			return SwPair(first , second)
		}
		
		fun <A:Any , B> acquireWeak(first:A , second:B):WeakReference<SwPair<A , B>>
		{
			return WeakReference(SwPair(first , second))
		}
	}
}
